src/curl_jolokia.c

```diff
@@ -0,0 +1,1045 @@
+#include "configfile.h"
+#include "plugin.h"
+#include "utils_complain.h"
+#include "utils_llist.h"
+
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <sys/un.h>
+
+#include <curl/curl.h>
+
+#include <yajl/yajl_parse.h>
+#if HAVE_YAJL_YAJL_VERSION_H
+#include <yajl/yajl_version.h>
+#endif
+
+#define CJO_DEFAULT_HOST "localhost"
+
+typedef size_t yajl_len_t;
+
+struct cjo_attribute_s /* {{{ */
+{
+	char *attribute_name;
+	char *attribute_match;
+	size_t attribute_match_len;
+	char *type;
+};
+typedef struct cjo_attribute_s cjo_attribute_t;
+/* }}} */
+
+struct cjo_bean_s /* {{{ */
+{
+	char *bean_name;
+	char *mbean_match;
+	char *mbean_namespace;
+	size_t mbean_match_len;
+	llist_t *attributes; /* list of cjo_attribute_t */
+};
+typedef struct cjo_bean_s cjo_bean_t;
+/* }}} */
+
+typedef enum { eNone = 0, eValue, eMBean } cjo_expect_token;
+
+/* {{{ */
+struct cjo_attribute_values_s {
+	const char *json_value;
+	yajl_len_t json_value_len;
+	const char *json_name;
+	yajl_len_t json_name_len;
+};
+typedef struct cjo_attribute_values_s cjo_attribute_values_t;
+/* }}} */
+
+struct cjo_membuffer_s {
+	char *buffer;
+	size_t size;
+	size_t used;
+};
+
+typedef struct cjo_membuffer_s cjo_membuffer_t;
+
+struct cjo_s /* {{{ */
+{
+	char *instance;
+	char *host;
+
+	char *sock;
+
+	char *url;
+	char *user;
+	char *pass;
+	char *credentials;
+
+	const cjo_bean_t *match_this_bean;
+	cjo_attribute_values_t *curr_attribute; /* points to the pool below. */
+	cjo_attribute_values_t *attributepool;
+	int attribute_pool_used;
+
+	const char *json_key;
+	yajl_len_t json_key_len;
+
+	cjo_expect_token expect;
+
+	_Bool verify_peer;
+	_Bool verify_host;
+	char *cacert;
+	struct curl_slist *headers;
+	char *post_body;
+	size_t post_body_len;
+	_Bool expect_escapes;
+	cdtime_t interval;
+	int timeout;
+
+	CURL *curl;
+	char curl_errbuf[CURL_ERROR_SIZE];
+	const unsigned char *start, *end;
+	cjo_membuffer_t replybuffer;
+	cjo_membuffer_t itembuffer;
+
+	size_t poolsize;
+	char *pool;
+	const char *poolmax;
+	char *pool_ptr;
+
+	yajl_handle yajl;
+	llist_t *bean_configs;
+	int max_attribute_count;
+	int max_value_len;
+};
+typedef struct cjo_s cjo_t; /* }}} */
+
+static void cjo_init_buffer(cjo_membuffer_t *buf, size_t initial_size) {
+	buf->buffer = calloc(1, initial_size);
+	buf->used = 0;
+	buf->size = initial_size;
+}
+
+static void cjo_append_buffer(cjo_membuffer_t *buf, const char *append,
+			      size_t appendlen) {
+	if (buf->used + appendlen + 1 > buf->size) {
+		size_t newsize = buf->size * 2;
+		char *newbuf = calloc(1, newsize);
+		memcpy(newbuf, buf->buffer, buf->used);
+		free(buf->buffer);
+		buf->buffer = newbuf;
+		buf->size = newsize;
+	}
+	memcpy(buf->buffer + buf->used, append, appendlen);
+	buf->used += appendlen;
+}
+
+static void cjo_release_buffercontent(cjo_membuffer_t *buf) {
+	sfree(buf->buffer);
+	buf->buffer = NULL;
+	buf->size = 0;
+	buf->used = 0;
+}
+
+static void cjo_remember_value(cjo_membuffer_t *ShouldBeInHere,
```

Comment:
This function frightens and confuses me. Here is what I *think* is going on:

The HTTP response is read into a `cjo_membuffer_t` and passed to this function as `ShouldBeInHere`. The JSON string pointed to by `ptr` is likely a pointer into this buffer. If not, then the value is copied into a scratch space, the `BufferHereIfNot`. The pointer `SetThisOne` is then updated to either point into the response buffer `ShouldBeInHere` or the scratch space `BufferHereIfNot`.

I think this is very complex and the benefit (reduced memory footprint?) doesn't justify the complexity. I'd much rather always allocate memory and store it in the pointer, for example:

```c
char *cjo_json_strdup(const unsigned char *s, yajl_len_t len) {
  char *out = calloc(1, len + 1);
  if (out == NULL)
    return NULL;
  memcpy(out, s, len);
  /* out[len] == 0 */
  return out;
}
/* … */
/* in cjo_cb_string */
sfree(attr->json_name);
attr->json_name = cjo_json_strdup(val, len);
```

Reply:
hm, the jolokia replies can become huge (up to megs) and this mechanism only catches the case that you have a gauge where the value was spread across two tcp chunks. Since the value is copied afterwards anyways, I'd like to leave it that way.

Reply:
Okay, but … we only need to copy attribute names and values, not the entire response. And that buffer is flushed when a map closes, presumably that's when an MBean has been read completely? Last but not least, the number of attributes being cached is capped at `MaxReadAttributes`.

How many attributes would you say an MBean has at most? Put another way, what you suggest users set `MaxReadAttributes` to?